cli/docker/container_manager.go

```diff
@@ -0,0 +1,545 @@
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	failedServices := make(map[string]error)
+	for serviceName := range services {
+		// Naming the container other than the service name, can cause issues with network calls
+		runningContainer, err := cm.runContainer(services[serviceName], containerCreateOptions{
+			name: serviceName,
+		}, envs)
+		if runningContainer != nil {
+			containers = append(containers, runningContainer)
+		}
+		if err != nil {
+			failedServices[serviceName] = err
+		}
+	}
+	// Even on failure we save the references to make sure containers will be cleaned up
+	cm.serviceContainers[workflowID] = append(cm.serviceContainers[workflowID], containers...)
+
+	if len(failedServices) != 0 {
+		errServices := fmt.Errorf("failed to start services")
+		for serviceName, err := range failedServices {
+			errServices = fmt.Errorf("%v: %w", errServices, err)
+			cm.logger.Errorf("Failed to start service container (%s): %s", serviceName, err)
+		}
+		return containers, errServices
+	}
+
+	for _, container := range containers {
+		if err := cm.healthCheckContainer(context.Background(), container); err != nil {
+			return containers, fmt.Errorf("container health check: %w", err)
+		}
+	}
+
+	return containers, nil
+}
+
+func (cm *ContainerManager) GetWorkflowContainer(workflowID string) *RunningContainer {
+	return cm.workflowContainers[workflowID]
+}
+
+func (cm *ContainerManager) GetServiceContainers(workflowID string) []*RunningContainer {
+	return cm.serviceContainers[workflowID]
+}
+
+func (cm *ContainerManager) DestroyAllContainers() error {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	cm.released = true
+	for _, container := range cm.workflowContainers {
+		cm.logger.Infof("ℹ️ Removing workflow container: %s", container.Name)
+		if err := container.Destroy(); err != nil {
+			return fmt.Errorf("destroy workflow container: %w", err)
+		}
+	}
+
+	for _, containers := range cm.serviceContainers {
+		for _, container := range containers {
+			if container == nil {
+				continue
+			}
+			cm.logger.Infof("Removing service container: %s", container.Name)
+			if err := container.Destroy(); err != nil {
+				return fmt.Errorf("destroy service container: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// We are not using the docker sdk for pull, start and create commands because:
+//   - We want to make sure the end user can easily debug using the same docker command we issue
+//     (hard to convert between sdk and cli api)
+//   - We'd like to support options generically,
+//     with the SDK we would need to parse the string ourselves to convert them properly to their own type
+//   - SDK differs from the CLI in some cases, for example pulling from private registry requires the exact token
+//     it can't automatically use the docker config
+func (cm *ContainerManager) runContainer(
+	container models.Container,
+	options containerCreateOptions,
+	envs map[string]string,
+) (*RunningContainer, error) {
+	if cm.released {
```

Comment:
Is this necessary? If the map of running containers is cleaned up when destroying, what else requires this released/not-released state?

Reply:
This is handling the edge case when the cleanup has started, but meanwhile a new workflow starts to run which creates a new container. With this we disable any future calls to the docker daemonl, making sure everything is cleaned up.

Reply:
Got it, thank you!